packages/next/client/route-loader.ts

```diff
@@ -292,13 +292,24 @@ function createRouteLoader(assetPrefix: string): RouteLoader {
     onEntrypoint(route: string, execute: () => unknown) {
       Promise.resolve(execute)
         .then((fn) => fn())
-        .then(
-          (exports: any) => ({
+        .then((exports: any) => {
+          const entrypoint = {
             component: (exports && exports.default) || exports,
             exports: exports,
-          }),
-          (err) => ({ error: err })
-        )
+          }
+          if (process.env.NODE_ENV !== 'production') {
+            const { isValidElementType } = require('react-is')
+            if (!isValidElementType(entrypoint.component)) {
+              throw markAssetError(
+                new Error(
+                  `The default export is not a React Component in page: "${route}"`
+                )
+              )
+            }
+          }
+          return entrypoint
+        })
+        .catch((error) => ({ error }))
```

Comment:
This is a change in behavior as we previously showed the error client-side without a hard navigation and moving the error here causes the route loading to be considered failed triggering a hard navigation to the page showing the server-side error instead (both in the overlay).  

Since tests passed, this behavior change might be alright if it seems preferable.